### PR TITLE
add tf recordio dataset op

### PR DESCRIPTION
**Pull request by yuyicg**
1. add a user defined tf recordio dataset op kernel.
2. add a python wrapper for the user defined kernel.

### CODE REVIEW

File: data/tensorflow/operator/recordio_dataset_op.cc [hunk 1]
```diff
@@ -0,0 +1,215 @@
+#include "tensorflow/core/framework/dataset.h"
+#include "tensorflow/core/framework/common_shape_fns.h"
+#include "../recordio/recordio_reader.h"
```
Comment (zou000) [inline data/tensorflow/operator/recordio_dataset_op.cc]: don't use relative path like .. if needed could add path to HEADER_DIR

File: data/tensorflow/operator/recordio_dataset_op.cc [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+                    "`chunk_count` must be >= 0 (0 means read all the chunks)"));
+
+    *output =
+        new Dataset(ctx, std::move(filenames), start_chunk, chunk_count);
+  }
+
+ private:
+  class Dataset : public GraphDatasetBase {
+   public:
+    explicit Dataset(OpKernelContext* ctx, std::vector<string> filenames,
```
Comment (zou000) [inline data/tensorflow/operator/recordio_dataset_op.cc]: pass filenames use const std::vector<string>&

File: data/tensorflow/operator/recordio_dataset_op.cc [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+class RecordIODatasetOp : public DatasetOpKernel {
+ public:
+  using DatasetOpKernel::DatasetOpKernel;
+
+  void MakeDataset(OpKernelContext* ctx, DatasetBase** output) override {
+    const Tensor* filenames_tensor;
+    OP_REQUIRES_OK(ctx, ctx->input("filenames", &filenames_tensor));
+    OP_REQUIRES(
+        ctx, filenames_tensor->dims() <= 1,
+        errors::InvalidArgument("`filenames` must be a scalar or a vector."));
```
Comment (zou000) [inline data/tensorflow/operator/recordio_dataset_op.cc]: 我们应该不需要filenames是vector的情况吧?filenames是vector 而又share start_chunk 和 chunk_count 的情况应该是不会发生

**Review by zou000 — COMMENTED** [review]
This is great progress. Left some comments.

File: data/tensorflow/operator/recordio_dataset_op.cc (hunk 1 above)
Comment (yuyicg) [reply data/tensorflow/operator/recordio_dataset_op.cc]: fixed.

File: data/tensorflow/operator/recordio_dataset_op.cc (hunk 2 above)
Comment (yuyicg) [reply data/tensorflow/operator/recordio_dataset_op.cc]: already change to const string& filename

File: data/tensorflow/operator/recordio_dataset_op.cc (hunk 3 above)
Comment (yuyicg) [reply data/tensorflow/operator/recordio_dataset_op.cc]: changed the first input param from filenames to scalar type filename.

File: data/tensorflow/operator/recordio_dataset_op.cc [hunk 4]
```diff
@@ -53,16 +48,16 @@ class RecordIODatasetOp : public DatasetOpKernel {
...
 
  private:
   class Dataset : public GraphDatasetBase {
    public:
-    explicit Dataset(OpKernelContext* ctx, std::vector<string> filenames,
+    explicit Dataset(OpKernelContext* ctx, const string& filename,
                      int64 start_chunk, int64 chunk_count)
         : GraphDatasetBase(ctx),
-          filenames_(std::move(filenames)),
+          filename_(std::move(filename)),
```
Comment (zou000) [inline data/tensorflow/operator/recordio_dataset_op.cc]: no need to use std::move

File: data/tensorflow/operator/recordio_dataset_op.cc [hunk 5]
```diff
@@ -125,24 +120,17 @@ class RecordIODatasetOp : public DatasetOpKernel {
             // We have reached the end of the current file, so maybe
             // move on to next file.
```
Comment (zou000) [inline data/tensorflow/operator/recordio_dataset_op.cc]: the comment is no long needed.

File: data/tensorflow/operator/recordio_dataset_op.cc [hunk 6]
```diff
@@ -171,19 +155,11 @@ class RecordIODatasetOp : public DatasetOpKernel {
...
-        TF_RETURN_IF_ERROR(env->NewRandomAccessFile(next_filename, &file_));
+        string filename = dataset()->filename_;
+        TF_RETURN_IF_ERROR(env->NewRandomAccessFile(filename, &file_));
 
         uint64 file_size = 0;
-        TF_RETURN_IF_ERROR(env->GetFileSize(next_filename, &file_size));
+        TF_RETURN_IF_ERROR(env->GetFileSize(filename, &file_size));
 
         reader_.reset(
             new io::RecordIOReader(file_.get(), file_size, dataset()->start_chunk_, dataset()->chunk_count_));
```
Comment (zou000) [inline data/tensorflow/operator/recordio_dataset_op.cc]: 现在我们的接口是(文件名,start_chunk,chunk_count) 这个在大文件时会有问题。每次产生dataset 时都需要读进整个文件建立index。我建议接口简单点就是（文件名，offset），只要从这个offset读一个chunk。

File: data/tensorflow/operator/recordio_dataset_op.cc (hunk 4 above)
Comment (yuyicg) [reply data/tensorflow/operator/recordio_dataset_op.cc]: fixed.

File: data/tensorflow/operator/recordio_dataset_op.cc (hunk 5 above)
Comment (yuyicg) [reply data/tensorflow/operator/recordio_dataset_op.cc]: fixed.

File: data/tensorflow/operator/recordio_dataset_op.cc (hunk 6 above)
Comment (yuyicg) [reply data/tensorflow/operator/recordio_dataset_op.cc]: 是有这个问题,大文件需要每次都遍历建索引,已经修改成仅包含filename和offset两个参数,且每次只读offset起始的一个chunk.

**Review by zou000 — APPROVED** [review]
LGTM!

**Pull request merged**